### PR TITLE
[core][state] List actors should show actors from dead jobs as well

### PR DESCRIPTION
**Pull request by rickyyx**
Signed-off-by: rickyyx <rickyx@anyscale.com>

## Why are these changes needed?
This is to enable use cases by the dashboard. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/state/state_head.py [hunk 1]
```diff
@@ -192,9 +192,16 @@ def _options_from_req(self, req: aiohttp.web.Request) -> ListApiOptions:
         timeout = int(req.query.get("timeout", 30))
         filters = self._get_filters_from_req(req)
         detail = convert_string_to_type(req.query.get("detail", False), bool)
+        show_dead_jobs = convert_string_to_type(
```
Comment (rkooo567) [inline dashboard/modules/state/state_head.py:195]: Let's just make it true for state API? I think we will sort the data to

alive actors -> dead actors 

before reporting it, so truncate should have the same effect all the time.

Comment (rickyyx) [reply dashboard/modules/state/state_head.py:195]: in that case, should we just set it True when querying GCS? And not make it configurable?

Comment (rkooo567) [reply dashboard/modules/state/state_head.py:195]: yes! I think there's no need to make it configurable (actually we should remove this flag. this is weird)

Comment (rickyyx) [reply dashboard/modules/state/state_head.py:195]: Do you mean remove this flag altogether and make it default to return dead jobs?

Comment (rkooo567) [reply dashboard/modules/state/state_head.py:195]: yes. OR we can just filter by states (like source-side filtering). I think this was the hack to reduce the returned # of actors

**Comment by rickyyx** [comment]
Let's try deprecate the `show_dead_jobs` flag from gcs maybe later?

I set it to True for state API though.

File: python/ray/tests/test_state_api_summary.py [hunk 2]
```diff
@@ -382,13 +382,14 @@ def verify():
...
         infeasible_summary = None
+        print(summary)
         for actor_class_name, s in summary.items():
             if ".Actor" in actor_class_name:
                 actor_summary = s
             elif ".Infeasible" in actor_class_name:
                 infeasible_summary = s
 
-        assert actor_summary["state_counts"]["PENDING_CREATION"] == 1
+        assert actor_summary["state_counts"]["DEPENDENCY_UNREADY"] == 1
```
Comment (rickyyx) [inline python/ray/tests/test_state_api_summary.py:392]: @rkooo567 I fail to figure out why the change in this PR has an impact on this task - do you have any ideas? 
Is the change expected if we are now getting actors from dead jobs by default? Might be some data mismatch from gcs data.

Comment (rkooo567) [reply python/ray/tests/test_state_api_summary.py:392]: Hmm this doesn't make sense at all? We don't have any actor that even has an dependency... we may need to dig deeper

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Request changes until we figure out the test failure

**Comment by rickyyx** [comment]
TODO

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 3]
```diff
@@ -649,8 +649,13 @@ Status GcsActorManager::CreateActor(const ray::rpc::CreateActorRequest &request,
       request.task_spec(), actor_namespace, actor_state_counter_);
   actor->UpdateState(rpc::ActorTableData::PENDING_CREATION);
   const auto &actor_table_data = actor->GetActorTableData();
-  // Pub this state for dashboard showing.
-  RAY_CHECK_OK(gcs_publisher_->PublishActor(actor_id, actor_table_data, nullptr));
+  RAY_CHECK_OK(gcs_table_storage_->ActorTable().Put(
```
Comment (rickyyx) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:652]: We are not persisting the actor state changes before...
So it makes a difference when listing actors including dead jobs, which takes data from the actor table rather than in the memory cache. Since we never update the table, actors that are stuck in "PENDING_CREATION" are never returned. 

Also, we are publishing the entire actor state, we could just do with the delta.

**Comment by rickyyx** [comment]
We need this PR for https://github.com/ray-project/ray/pull/33555 since we will need to list actors to get the "repr_name" when summarizing tasks including actor tasks.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc (hunk 3 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:649]: good catch. but this also means we should now handle the race condition. E.g., before the callback is executed if DestroyActor is called (e.g., HandleKillActor request comes in), we can have problems. So we should have the logic like this

```
  auto it = registered_actors_.find(actor_id);
  if (it == registered_actors_.end()) {
    RAY_LOG(INFO) << "Actor is already destroyed " << actor_id;
    return;
  }
```

I think it is safe to just return if it is already deleted.

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:649]: Can you also add a unit test for this case?

**Review by rkooo567 — CHANGES_REQUESTED** [review]
One potential race condition comment.

**Comment by rkooo567** [comment]
also let's make sure to run stress_test_dead_actors before merging the PR

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc (hunk 3 above)
Comment (rickyyx) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:649]: Are you saying we shouldn't publish actor if the actor is removed from registered actors due to a concurrent actor kill?

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:649]: not just publish, but we shouldn't also do scheduling itself (since it is already killed)

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:649]: yeah that's correct!

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:649]: and actually I think the rest of methods should be inside the callback

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:649]:
```
  RemoveUnresolvedActor(actor);

  // Update the registered actor as its creation task specification may have changed due
  // to resolved dependencies.
  registered_actors_[actor_id] = actor;
  // Schedule the actor.
  gcs_actor_scheduler_->Schedule(actor);
```

this part

**Comment by rkooo567** [comment]
Instead of handling the persistency issue, we decided to just use cache directly
